### PR TITLE
[POC] - allowing module developers to insert form element to any position we like in the form.

### PR DESCRIPTION
**Pull request by tomas862**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | By using css flex and order properties for each element we can control in which position elements are placed. So if an element from the module appears below every other elements we can assign order: 0 to module form element and it will appear in the top of all elements in flex container.
| Type?         | new feature
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/13876
| How to test?  | Currently its Proof of concept - not testing information available right now

[POC] - **PROOF OF CONCEPT** . The code is not suppose to be working. Some parts are only theories as this PR purpose is just to explain all parts involved in making this feature to work. If we agree for this when I'll remove POC and implement everything

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/14094)

### CODE REVIEW

File: src/PrestaShopBundle/Resources/views/Admin/macros.html.twig [hunk 1]
```diff
@@ -254,7 +254,7 @@
 
   {% set class = extraVars.class|default('') %}
 
-  <div class="form-group row {{ class }}">
+  <div class="form-group row {{ class }}" {% if extraVars.order is defined %} style="order: {{ extraVars.order }}; -webkit-order: {{ extraVars.order }}" {% endif %}>
```
Comment (tomas862) [inline src/PrestaShopBundle/Resources/views/Admin/macros.html.twig:256]: css in html - unfortunately  this part is unavoidable. 

Another option could be:
Unless we can create classes for each order posiiton but we must know how many elements exist in the first place. We can generate **100** classes - but again if we hhave more then 100 elements then its a problem with this functionality.

**Comment by tomas862** [comment]
and in module:

1 simple option:
```php
/** @var FormBuilderInterface $formBuilder */
        $formBuilder = $params['form_builder'];
        $formBuilder->add('is_allowed_for_review', SwitchType::class, [
            'label' => $this->getTranslator()->trans('Allow reviews', [], 'Modules.Ps_DemoCQRSHooksUsage'),
            'required' => false,
            'order' => 0,
        ]);

```

2 harder option:

```php
/** @var CustomPrestaShopFormBuilderInterface $formBuilder */
        $formBuilder = $params['form_builder'];
        $formBuilder->addAfter('another_element_in_form', 'is_allowed_for_review', SwitchType::class, [
            'label' => $this->getTranslator()->trans('Allow reviews', [], 'Modules.Ps_DemoCQRSHooksUsage'),
            'required' => false,
        ]);
```

**Comment by mickaelandrieu** [comment]
Hello,

I have the feeling that I repeat myself: this is the wrong solution to the wrong problem **but** it proves there is a developer experience issue here :+1: 

If we fix our form theme, all our template will look like this:

```
"" some static content """
{{ form(form) }}
""" some other static content"""
```

Once you do that, the only thing you'll have to re-order the form using an event (and this could be interesting) to control the order of the form fields at the rendering time! But for now, as each form field is rendered in the template, I don't see a maintainable solution to address your specific need.

I don't think a solution based on CSS, that hides the real position of each element, is something we should implement in the core.

**Comment by tomas862** [comment]
Hi, sure if this 

```twig
{{ form(form) }}
```
is going to be implemented no css magic has to be applied - have to agree with that. :+1:

**Comment by matks** [comment]
> If we fix our form theme, all our template will look like this:
> 
> ```
> "" some static content """
> {{ form(form) }}
> """ some other static content"""
> ```

I dont think we'll be able to reach this point because our designer Tristan aims for a very smooth UX which cannot be obtained with a default Form Theme used everywhere in the application.

> Once you do that, the only thing you'll have to re-order the form using an event (and this could be interesting) to control the order of the form fields at the rendering time!

Don't you think it's a bit overkill to use a hook only to modify the order of the fields ?

> I don't think a solution based on CSS, that hides the real position of each element, is something we should implement in the core.

It brings a great benefits: a theme developer or integrator can obtain full control of the position of each elements without having to deal with php or hooks

**Pull request closed without merging**